File: 0001-two-sum/solution.py
# https://leetcode.com/problems/two-sum/solutions/737092/sum-megapost-python3-solution-with-a-detailed-explanation/
# 0001-two-sum

class Solution:
    def twoSum(self, nums: List[int], target: int) -> List[int]:
        return self.solution2(nums, target)

    def solution(self, nums: List[int], target: int) -> List[int]:
        for i in range(len(nums)):
            for j in range(i + 1, len(nums)):
                if nums[i] + nums[j] == target:
                    return [i, j]

    def solution2(self, nums: List[int], target: int) -> List[int]:
        indexDict = {}
        
        for i in range(len(nums)):
            otherIndexForSum = indexDict.get(target - nums[i])
            if otherIndexForSum is not None:
                return [i, otherIndexForSum]
            indexDict[nums[i]] = i

File: 0001-two-sum/solution.ts
function twoSum(nums: number[], target: number): [number, number] {
  const dict: Record<number, number> = {};
  for(let i = 0; i < nums.length; i++) {
    const compliment = target - nums[i];
    if(dict[compliment] != null) {
      return [i, dict[compliment]];
    }
    dict[nums[i]] = i;
  }
};

// function twoSum(nums: number[], target: number): [number, number] {
//   const numsWithIndex= nums.map((num, i) => [num, i]);
//   numsWithIndex.sort((a, b) => a[0] - b[0]);
//   let i = 0;
//   let j = nums.length - 1;
//   while (i < j) {
//     const sum = numsWithIndex[i][0] + numsWithIndex[j][0];
//     if(sum > target) {
//       j--;
//       continue;
//     }
//     if(sum < target) {
//       i++;
//       continue;
//     }
//     return [numsWithIndex[i][1], numsWithIndex[j][1]];
//   }
// };
